cmd/start.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2017 the lostromos Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	homedir "github.com/mitchellh/go-homedir"
+	"github.com/spf13/cobra"
+	"github.com/wpengine/lostromos/crwatcher"
+	"github.com/wpengine/lostromos/printctlr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+var (
+	crdName      string
+	crdGroup     string
+	crdVersion   string
+	crdNamespace string
+	kubeFile     string
+	kubeMaster   string
+	inKube       bool
+)
+
+var startCmd = &cobra.Command{
+	Use:   "start",
+	Short: `Start the server.`,
+	Run: func(command *cobra.Command, args []string) {
+		startServer()
+	},
+}
+
+func init() {
+	LostromosCmd.AddCommand(startCmd)
+	startCmd.Flags().StringVar(&kubeMaster, "kube-master", "", "address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
+	startCmd.Flags().StringVar(&kubeFile, "kube-config", filepath.Join(homeDir(), ".kube", "config"), "absolute path to the kubeconfig file. Only required if out-of-cluster.")
+	startCmd.Flags().BoolVar(&inKube, "k8s", false, "When set uses the K8s service account token. Use when running in a cluter.")
+	startCmd.Flags().StringVar(&crdName, "crd-name", "", "the plural name of the CRD you want monitored (ex: users)")
+	startCmd.Flags().StringVar(&crdGroup, "crd-group", "", "the group of the CRD you want monitored (ex: stable.wpengine.io)")
+	startCmd.Flags().StringVar(&crdVersion, "crd-version", "v1", "the group of the CRD you want monitored")
+	startCmd.Flags().StringVar(&crdNamespace, "crd-namespace", metav1.NamespaceNone, "(optional) the namespace of the CRD you want monitored, only needed for namespaced CRDs (ex: default)")
+}
+
+func homeDir() string {
+	home, err := homedir.Dir()
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+	return home
+}
+
+func getKubeClient() *restclient.Config {
+	var (
+		cfg *restclient.Config
+		err error
+	)
+	if inKube {
+		cfg, err = restclient.InClusterConfig()
+	} else {
+		cfg, err = clientcmd.BuildConfigFromFlags(kubeMaster, kubeFile)
+	}
+	if err != nil {
+		panic(err.Error())
+	}
+	return cfg
+}
+
+func buildCRWatcher(cfg *restclient.Config) *crwatcher.CRWatcher {
```

Comment:
Do we need a struct here or should we be adding these things to the viper configuration?

Reply:
The config struct? We can't just leverage the viper configuration here as we want this to be a library that could be imported into other code bases. Those code bases shouldn't have to use viper just to use our code. That is our implementation detail not theirs.